File: checkin_frontend/lib/core/models/task/task_public_detail_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../subtask_instance/subtask_combined_list_model.dart';
import '../enums/task_enums.dart';

part 'task_public_detail_model.freezed.dart';
part 'task_public_detail_model.g.dart';

@freezed
sealed class TaskPublicDetailModel with _$TaskPublicDetailModel {
  const factory TaskPublicDetailModel({
    String? id,
    String? hash,
    @Default('') String title,
    String? notes,
    DateTime? deadLine,
    @Default(TaskState.inProgress) TaskState state,
    @Default(SubtaskMode.individual) SubtaskMode subtaskMode,
    @Default(false) bool requiresAuthenticationToComplete,
    @Default([]) List<SubtaskCombinedListModel> subtasks,
    String? allowedDomain,
    @Default(false) bool isForbidden,
    String? forbiddenMessage,
  }) = _TaskPublicDetailModel;

  factory TaskPublicDetailModel.fromJson(Map<String, dynamic> json) =>
      _$TaskPublicDetailModelFromJson(json);
}

File: CheckIn.Api.Common/Models/Details/TaskPublicDetailModel.cs
using CheckIn.Api.Common.Enums;
using CheckIn.Api.Common.Models.Interfaces;
using CheckIn.Api.Common.Models.Lists;
using TaskStatus = System.Threading.Tasks.TaskStatus;

namespace CheckIn.Api.Common.Models.Details;

public record TaskPublicDetailModel : IEntityModel
{
    public Guid Id { get; init; }
    public string Hash { get; set; } = null!;
    public required string Title { get; init; }
    public string? Notes { get; init; }
    public DateTime DeadLine { get; init; }
    public TaskState State { get; init; }
    public SubtaskMode SubtaskMode { get; init; }
    public bool RequiresAuthenticationToComplete { get; init; }
    public string? AllowedDomain { get; init; }
    public List<SubtaskCombinedListModel> Subtasks { get; init; } = new();
    public bool IsForbidden { get; set; }
    public string? ForbiddenMessage { get; set; }
}